### PR TITLE
Switch the telemetry to new futures

### PR DESCRIPTION
**Pull request by tomaka**
cc #3099 

The majority of the switch to new futures is straight-forward, but for the telemetry it is pretty much not straight-forward.
Since libp2p is using futures v0.1, and the telemetry has a wrapper that guarantees that one telemetry write equals one WebSocket frame, we need to have a compatibility shim.

After this PR, the API of `substrate-telemetry` uses new futures.

### CODE REVIEW

File: core/telemetry/src/worker/node.rs [hunk 1]
```diff
@@ -169,36 +168,47 @@ where TTrans: Clone, TTrans::Output: Sink<SinkItem = BytesMut, SinkError = TSink
...
-			} else if self.need_flush && self.sink.poll_complete()?.is_ready() {
-				self.need_flush = false;
+				let item_len = item.len();
+				if let Err(err) = Sink::start_send(Pin::new(&mut self.sink), item) {
+					return Poll::Ready(Err(err))
+				}
+				trace!(target: "telemetry", "Successfully sent {:?} bytes message to {}",
+					item_len, my_addr);
+				self.need_flush = true;
+
```
Comment (bkchr) [inline core/telemetry/src/worker/node.rs]:
```suggestion
```

File: core/telemetry/src/worker/node.rs [hunk 2]
```diff
@@ -169,36 +168,47 @@ where TTrans: Clone, TTrans::Output: Sink<SinkItem = BytesMut, SinkError = TSink
...
 				}
 
-			} else if self.need_flush && self.sink.poll_complete()?.is_ready() {
-				self.need_flush = false;
+				let item_len = item.len();
+				if let Err(err) = Sink::start_send(Pin::new(&mut self.sink), item) {
+					return Poll::Ready(Err(err))
+				}
+				trace!(target: "telemetry", "Successfully sent {:?} bytes message to {}",
+					item_len, my_addr);
```
Comment (bkchr) [inline core/telemetry/src/worker/node.rs]:
```suggestion
					item_len, my_addr
				);
```

File: core/telemetry/src/worker/node.rs [hunk 3]
```diff
@@ -169,36 +168,47 @@ where TTrans: Clone, TTrans::Output: Sink<SinkItem = BytesMut, SinkError = TSink
...
+					return Poll::Pending
 				}
 
-			} else if self.need_flush && self.sink.poll_complete()?.is_ready() {
-				self.need_flush = false;
+				let item_len = item.len();
+				if let Err(err) = Sink::start_send(Pin::new(&mut self.sink), item) {
+					return Poll::Ready(Err(err))
+				}
+				trace!(target: "telemetry", "Successfully sent {:?} bytes message to {}",
```
Comment (bkchr) [inline core/telemetry/src/worker/node.rs]:
```suggestion
				trace!(
					target: "telemetry", "Successfully sent {:?} bytes message to {}",
```

File: core/telemetry/src/worker.rs [hunk 4]
```diff
@@ -58,19 +57,29 @@ pub struct TelemetryWorker {
 /// The pile of libp2p transports.
 #[cfg(not(target_os = "unknown"))]
 type WsTrans = libp2p::core::transport::timeout::TransportTimeout<
-	libp2p::core::transport::OrTransport<
-		libp2p::core::transport::map::Map<
-			OptionalTransport<wasm_ext::ExtTransport>,
-			fn(wasm_ext::Connection, ConnectedPoint) -> StreamSink<wasm_ext::Connection>
+	libp2p::core::transport::map::Map<
```
[the remark this answers is not in the document]
Comment (tomaka) [reply core/telemetry/src/worker.rs]: The problem here is that I'm using libp2p in a way it wasn't really designed for, but theoretically I could add some utility that allows using a `Box` here instead.

[the remark this answers is not in the document]
Comment (bkchr) [reply core/telemetry/src/worker.rs]: It is okay, probably more rust faults of not having `impl Something`.

**Pull request merged**